drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,344 @@
+			fifo_ths_cfg = &read_config->triggers[i];
+			continue;
+		}
+
+		if (read_config->triggers[i].trigger == SENSOR_TRIG_FIFO_FULL) {
+			fifo_full_cfg = &read_config->triggers[i];
+			continue;
+		}
+	}
+
+	/* fill fifo h/w status */
+	fifo_th = (lsm6dsv16x->fifo_status[1] & 0x80) ? 1 : 0;
+	fifo_full = (lsm6dsv16x->fifo_status[1] & 0x20) ? 1 : 0;
+	fifo_count = (uint16_t)lsm6dsv16x->fifo_status[1] & 0x1U;
+	fifo_count = (fifo_count << 8) | lsm6dsv16x->fifo_status[0];
+	lsm6dsv16x->fifo_count = fifo_count;
+
+	bool has_fifo_ths_trig = fifo_ths_cfg != NULL && fifo_th == 1;
+	bool has_fifo_full_trig = fifo_full_cfg != NULL && fifo_full == 1;
+
+	/* check for spurious interrupts */
+	if (!has_fifo_ths_trig && !has_fifo_full_trig) {
+		lsm6dsv16x->streaming_sqe = NULL;
+		gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	struct rtio_cqe *cqe;
+	int res = 0;
+
+	do {
+		cqe = rtio_cqe_consume(lsm6dsv16x->rtio_ctx);
+		if (cqe != NULL) {
+			if ((cqe->result < 0) && (res == 0)) {
+				LOG_ERR("Bus error: %d", cqe->result);
+				res = cqe->result;
+			}
+			rtio_cqe_release(lsm6dsv16x->rtio_ctx, cqe);
+		}
+	} while (cqe != NULL);
+
+	/* Bail/cancel attempt to read sensor on any error */
+	if (res != 0) {
+		rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, res);
+		lsm6dsv16x->streaming_sqe = NULL;
+		return;
+	}
+
+	enum sensor_stream_data_opt data_opt;
+
+	if (has_fifo_ths_trig && !has_fifo_full_trig) {
+		/* Only care about fifo threshold */
+		data_opt = fifo_ths_cfg->opt;
+	} else if (!has_fifo_ths_trig && has_fifo_full_trig) {
+		/* Only care about fifo full */
+		data_opt = fifo_full_cfg->opt;
+	} else {
+		/* Both fifo threshold and full */
+		data_opt = MIN(fifo_ths_cfg->opt, fifo_full_cfg->opt);
+	}
+
+	if (data_opt == SENSOR_STREAM_DATA_NOP || data_opt == SENSOR_STREAM_DATA_DROP) {
+		uint8_t *buf;
+		uint32_t buf_len;
+
+		/* Clear streaming_sqe since we're done with the call */
+		if (rtio_sqe_rx_buf(lsm6dsv16x->streaming_sqe, sizeof(struct lsm6dsv16x_fifo_data),
+				    sizeof(struct lsm6dsv16x_fifo_data), &buf, &buf_len) != 0) {
+			rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, -ENOMEM);
+			lsm6dsv16x->streaming_sqe = NULL;
+			gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+			return;
+		}
+
+		struct lsm6dsv16x_fifo_data *rx_data = (struct lsm6dsv16x_fifo_data *)buf;
+
+		memset(buf, 0, buf_len);
+		rx_data->header.is_fifo = 1;
+		rx_data->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+		rx_data->int_status = lsm6dsv16x->fifo_status[1];
+		rx_data->fifo_count = 0;
+		rtio_iodev_sqe_ok(lsm6dsv16x->streaming_sqe, 0);
+		lsm6dsv16x->streaming_sqe = NULL;
+
+		if (data_opt == SENSOR_STREAM_DATA_DROP) {
+			const struct lsm6dsv16x_config *config = dev->config;
+			stmdev_ctx_t *ctx = (stmdev_ctx_t *)&config->ctx;
+
+			/* Flush the FIFO by disabling it. */
+			lsm6dsv16x_fifo_mode_set(ctx, LSM6DSV16X_BYPASS_MODE);
```

Comment:
Keep in mind you are likely in an ISR here so if this does a blocking bus request then you will be blocking in the ISR, instead an additional bus operation needs to be asynchronously started at this point before returning a completion to the submission.

Reply:
Correct, I missed that we may be in ISR context (I guess that sometimes we may be also in thread context, but even so this is another point). So, I'll use an asynchronous direct bus write operation to act as the .lsm6dsv16x_fifo_mode_set().